colossalai/nn/optimizer/distributed_came.py

```diff
@@ -0,0 +1,284 @@
+        self.working_shape[id(param)]
+        # return tensor.reshape(*working_shape)
+        ori_shape = self.ori_shape[id(param)]
+        if not (len(ori_shape) >= 2 and len(param.grad.data.shape) == 1):
+            return tensor
+        if self.gather_before_compute[id(param)]:
+            rank = dist.get_rank(group=self.zero_parallel_group)
+            length = param.data.shape[0]
+            return torch.flatten(tensor)[rank * length : (rank + 1) * length]
+        return torch.flatten(tensor)
+
+    def step(self, closure=None):
+        """Performs a single optimization step.
+        Args:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+
+        for group in self.param_groups:
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+                grad = self._unflatten_grad_tensor_by_param(p) if self.zero else p.grad.data
+                if grad.dtype in {torch.float16, torch.bfloat16}:
+                    grad = grad.float()
+                if grad.is_sparse:
+                    raise RuntimeError("CAME does not support sparse gradients.")
+
+                state = self.state[p]
+                # zero下grad_shape是原始grad经过flatten后再切割(只有一维)
+                grad_shape = grad.shape
+
+                factored = self._get_options(grad_shape)
+                # State Initialization
+                if len(state) == 0:
+                    state["step"] = 0
+
+                    state["exp_avg"] = torch.zeros_like(grad)
+                    if factored:
+                        state["exp_avg_sq_row"] = torch.zeros(grad_shape[:-1]).type_as(grad)
+                        state["exp_avg_sq_col"] = torch.zeros(grad_shape[:-2] + grad_shape[-1:]).type_as(grad)
+
+                        state["exp_avg_res_row"] = torch.zeros(grad_shape[:-1]).type_as(grad)
+                        state["exp_avg_res_col"] = torch.zeros(grad_shape[:-2] + grad_shape[-1:]).type_as(grad)
+                    else:
+                        state["exp_avg_sq"] = torch.zeros_like(grad)
+
+                    state["RMS"] = 0
+
+                state["step"] += 1
+                state["RMS"] = self._rms(p.data, p)
+
+                update = (grad**2) + group["eps"][0]
+                if factored:
+                    exp_avg_sq_row = state["exp_avg_sq_row"]
+                    exp_avg_sq_col = state["exp_avg_sq_col"]
+
+                    # 局部平均
+                    sq_mean_row = update.mean(dim=-1)
+                    sq_mean_col = update.mean(dim=-2)
+                    if self.distributed:
+                        # 全局同步
+                        if self.clip_method[id(p)] == "row":
+                            dist.all_reduce(sq_mean_row, op=dist.ReduceOp.AVG, group=self.tensor_parallel_group)
```

Comment:
![image](https://github.com/hpcaitech/ColossalAI/assets/23111350/de2b4be3-9b7a-42ce-91ef-0ed5758ad999)

For better compatibility, SUM may be better